catalog/core/catalog-core-api-impl/src/main/java/ddf/catalog/util/impl/Poller.java

```diff
@@ -0,0 +1,270 @@
+        new ExecutorCompletionService<>(pollTimeoutWatcherThreadPool);
+    current.forEach(
+        (key, loader) -> {
+          try {
+            // Start the process to load a new value
+            final Future<V> future = pollThreadPool.submit(loader);
+
+            // Submit a separate process to wait for the load to finish or timeout
+            completionService.submit(
+                () -> getCommitRunnable(key, future, timeout, timeoutTimeUnit));
+          } catch (RejectedExecutionException e) {
+            throw new IllegalStateException("Unable to poll " + key, e);
+          }
+        });
+
+    // In order of loader process completion, commit the new value to the cache or handles the
+    // timeout or RuntimeException
+    for (int i = 0; i < currentCount; i++) {
+      final Runnable commitRunnable;
+      try {
+        commitRunnable = completionService.take().get();
+      } catch (CancellationException | ExecutionException e) {
+        throw new IllegalStateException(
+            "Exception while waiting for polling to complete or time out", e);
+      } catch (InterruptedException e) {
+        Thread.currentThread().interrupt();
+        throw new IllegalStateException(
+            "Interrupted while waiting for polling to complete or time out", e);
+      }
+
+      commitRunnable.run();
+    }
+
+    LOGGER.trace("Successfully updated the cache for each of the {} current", currentCount);
+  }
+
+  /** Removes keys in the cache that are not in the {@link Set} of current keys */
+  private void removeNoncurrentKeysFromTheCache(final Set<K> currentKeys) {
+    final Set<K> keysInTheCache = cache.asMap().keySet();
+    final Set<K> keysToRemoveFromTheCache = Sets.difference(keysInTheCache, currentKeys);
+    if (!keysToRemoveFromTheCache.isEmpty()) {
+      if (LOGGER.isDebugEnabled()) {
+        LOGGER.debug(
+            "Found {} entries to remove from the cache: {}",
+            keysToRemoveFromTheCache.size(),
+            StringUtils.join(keysToRemoveFromTheCache, ", "));
+      }
+      keysToRemoveFromTheCache
+          .stream()
+          .peek(key -> LOGGER.debug("Removing {} from the cache", key))
+          .forEach(cache::invalidate);
+    } else {
+      LOGGER.trace("Found no entries to remove from the cache");
+    }
+  }
+
+  /**
+   * Waits at most the given timeout for the loader process to complete
+   *
+   * @return a {@link Runnable} to commit the new value or handle the timeout or {@link
+   *     RuntimeException}
+   */
+  private Runnable getCommitRunnable(
+      final K key, final Future<V> loaderFuture, long timeout, TimeUnit timeoutTimeUnit) {
+    try {
+      final V newValue = loaderFuture.get(timeout, timeoutTimeUnit);
+      if (newValue == null) {
+        throw new IllegalStateException("The new value for " + key + " may not be null");
+      }
+
+      return () -> cacheNewValue(key, newValue);
+    } catch (TimeoutException e) {
+      loaderFuture.cancel(true);
```

Comment:
I don't think Future#get(long, TimeUnit) cancels the thread itself. Shouldn't that task manually be cancelled to save resources because the result will not be used?